### ISSUE TITLE
--restrict-lexicon

### ISSUE THREAD

**Issue by wangpeng138375** [awslabs/sockeye#867]
I find `--restrict-lexicon` at [arguments.py](https://github.com/awslabs/sockeye/blob/master/sockeye/arguments.py). Has this argument beem implemented and is there any instructions for it?

**Comment by fhieber** [comment]
There is some documentation about this here (https://github.com/awslabs/sockeye/tree/master/sockeye_contrib/fast_align), describing how to create a lexicon from fast_align models. If used in decoding, the model can only generate words that are part of this lexicon.

**Comment by wangpeng138375** [comment]
Thanks, and I close this issue.

**Comment by wangpeng138375** [comment]
When I got the result of lex.out from fast_align, I deleted most of the lines, leaving only some words that I focused on.  The model exactly only generated words that are part of this lexicon in decoding, as you said above. But the model can never gererate words that I deleted.

I have two questions:

1> As the results show, it seems that it is not the combination of discrete probability and continuous probability, but only discrete probability is used. Why can't words out of the lexicon be generated?
2> I added `--restrict-lexicon key1:path1 key2:path2 ...` when starting the model, and used ` {"text": "some input string", "restrict_lexicon": "key"}` when tanslating a sentence.  Is there any way to dynamically add any vocabularies when translating?  Because I want the same source word to be translated into different target words in different sentences depending on different lexicon.

**Comment by bricksdont** [comment]
1)

> As the results show, it seems that it is not the combination of discrete probability and continuous probability, but only discrete probability is used.

I don't understand what you mean by a "combination of discrete and continuous probability". Could you perhaps explain this a bit more?

> Why can't words out of the lexicon be generated?

Because that is precisely what `--restrict-lexicon` does: it limits possible translations to the target tokens present in a lexicon.

What is perhaps confusing is that for a given source sentence, the list of allowed target tokens that can be generated is [the set of all target tokens that appear in the lexicon as the translation of at least one of the tokens in the source sentence](https://github.com/awslabs/sockeye/blob/master/sockeye/lexicon.py#L187). Only this subset of target vocabulary entries is used to make predictions, in the sense that the softmax used to create a distribution over the target vocabulary is fed only the probabilities of this subset of the vocabulary.

You perhaps assumed that if you delete certain source tokens from the lexicons, the model is free to choose a translation for them among all possible target words. But there is no token-level alignment in a standard NMT model, it cannot be determined which source token "causes" which target token.

3) The paper being referenced is: https://arxiv.org/pdf/1705.01991.pdf

**Comment by wangpeng138375** [comment]
Thanks from you detail reply.
Maybe it’s appropriate to call it "combination of probability coming from discrete mapping and probability coming from continuous representation (the NMT probability)" instead of "combination of discrete and continuous probability". That come from this paper [Incorporating Discrete Translation Lexicons into Neural Machine Translation](https://arxiv.org/pdf/1606.02006.pdf). **Has sockey2 implemented this paper?** I find only one [function](https://github.com/awslabs/sockeye/blob/master/sockeye/lexicon.py#L59) quotes this paper and one [issues](https://github.com/awslabs/sockeye/issues/323) mentions it. 

My 1> question was that if I delete certain source tokens from the lexicons, the model will use NMT probability only , instead of the combination of two.

**Comment by fhieber** [comment]
Sockeye2 has not implemented this paper. It was part of an old version of Sockeye 1 (as the issue you link to states), but was removed in the meantime.
The current `--restrict-lexicon` functionality simply causes the final output layer to be defined over a subset of tokens (the ones that are part of the lexicon). There is no combination of probabilities at all, in fact the lexicon that you would provide, does not even store any probabilities, it simply stores source to target word mappings that should define the set of target words the beam search during inference can search over.

**Comment by wangpeng138375** [comment]
Could you tell me the reason why removed this implementation, is it not as good as claimed in the paper? And is there any plan to reimplement it in the future?

**Comment by fhieber** [comment]
I honestly don't remember, this is a few years back. From our side there are no plans to add this back, but feel free to open a PR.

**Comment by wangpeng138375** [comment]
Thank you very much，I close it.

**Comment by mjdenkowski** [comment]
Using word alignment probabilities introduces a bias towards pre-trained word level mappings.  This was a great way to boost early NMT models that had some known weaknesses and didn't learn as well as modern models.  As NMT models got stronger, we passed a point where adding a word alignment bias actually degraded quality compared to letting the model learn everything on its own.

**Comment by wangpeng138375** [comment]
@mjdenkowski Thank you for your explanation. My original idea is to use sockeye to do paraphrase generation(monolingual), and I want to use the idea of this paper to integrate synonyms,  which I think maybe achieve the goal of diverse paraphrase. Anyway, thank you for your explanation. I'll look for some other ways.
